rust-query-crlite/src/main.rs

```diff
@@ -0,0 +1,794 @@
+//! # rust-query-crlite
+//!
+extern crate base64;
+extern crate bincode;
+extern crate byteorder;
+extern crate clap;
+extern crate der_parser;
+extern crate hex;
+extern crate log;
+extern crate num_bigint;
+extern crate pem;
+extern crate reqwest;
+extern crate rust_cascade;
+extern crate rustls;
+extern crate serde;
+extern crate sha2;
+extern crate stderrlog;
+extern crate x509_parser;
+
+use byteorder::{LittleEndian, ReadBytesExt};
+use clap::Parser;
+use der_parser::oid;
+use log::*;
+use rust_cascade::Cascade;
+use serde::Deserialize;
+use sha2::{Digest, Sha256};
+use std::collections::{HashMap, HashSet};
+use std::fmt::Display;
+use std::io::prelude::Write;
+use std::io::{BufReader, Read};
+use std::net::{TcpStream, ToSocketAddrs};
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+use x509_parser::prelude::*;
+
+const ICA_LIST_URL: &str =
+    "https://ccadb-public.secure.force.com/mozilla/MozillaIntermediateCertsCSVReport";
+
+const STAGE_ATTACH_URL: &str = "https://firefox-settings-attachments-cdn.stage.mozaws.net/";
+const STAGE_URL: &str = "https://settings-cdn.stage.mozaws.net/v1/buckets/security-state-staging/collections/cert-revocations/records";
+
+const PROD_ATTACH_URL: &str = "https://firefox-settings-attachments.cdn.mozilla.net/";
+const PROD_URL: &str = "https://firefox.settings.services.mozilla.com/v1/buckets/security-state/collections/cert-revocations/records";
+
+const COVERAGE_SERIALIZATION_VERSION: u8 = 1;
+const COVERAGE_V1_ENTRY_BYTES: usize = 48;
+
+const ENROLLMENT_SERIALIZATION_VERSION: u8 = 1;
+const ENROLLMENT_V1_ENTRY_BYTES: usize = 32;
+
+const OID_SCT_EXTENSION: &der_parser::Oid = &oid!(1.3.6 .1 .4 .1 .11129 .2 .4 .2);
+
+type CRLiteKey = Vec<u8>;
+type DERCert = Vec<u8>;
+type EnrollmentKey = Vec<u8>;
+type IssuerDN = Vec<u8>;
+type IssuerSPKIHash = Vec<u8>;
+type LogID = Vec<u8>;
+
+fn crlite_key(issuer_spki_hash: &[u8], serial: &[u8]) -> CRLiteKey {
+    let mut key = issuer_spki_hash.to_vec();
+    key.extend_from_slice(serial);
+    key
+}
+
+fn enrollment_key(issuer_dn: &[u8], issuer_spki: &[u8]) -> EnrollmentKey {
+    let mut hasher = Sha256::new();
+    hasher.update(issuer_dn);
+    hasher.update(issuer_spki);
+    hasher.finalize().to_vec()
+}
+
+// x509_parser parses the serial number to a BigUint, but DER integers are signed
+// so we need to retain a leading 0 byte in some cases where BitUint::to_bytes_be
+// drops it.
+fn serial_bytes(n: &num_bigint::BigUint) -> Vec<u8> {
```

Comment:
We've had problems in the past with improperly encoded serial numbers (e.g. a leading 0 byte should be present but isn't). Is there a way to directly use the encoded bytes of the serial number (minus the leading integer tag and length byte(s))?